Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,249 @@
+  `Array.size`
+-/
+def size (_ : Vector α n) := n
+
+/--
+  `Vector.nil` produces an empty vector
+-/
+@[match_pattern]
+def nil : Vector α 0 := ⟨Array.empty, by rfl⟩
+
+/--
+  `Vector.cons` is the list cons pattern for vectors
+-/
+@[match_pattern]
+def cons (x : α) (xs : Vector α n) : Vector α (Nat.succ n) :=
+  match xs with
+  | ⟨{data := xs'}, h⟩ => ⟨{data := x :: xs'}, congrArg Nat.succ h⟩
+
+/--
+  `Vector.head` produces the head of a vector
+-/
+def head (v : Vector α (Nat.succ n)) :=
+  match v with
+  | ⟨{data := a :: _}, _⟩ => a
+
+
+theorem head_cons (a : α) : ∀ v : Vector α n, head (cons a v) = a := by
+  intro v
+  match v with
+  | ⟨{data := xs}, h⟩ => rfl
+
+/--
+  `Vector.tail` produces the tail of a vector
+-/
+def tail (v : Vector α n) : Vector α (n-1):=
+  match v with
+  | ⟨{data := []}, h⟩ => ⟨{data := []}, congrArg Nat.pred h⟩
+  | ⟨{data := _ :: xs}, h⟩ => ⟨{data := xs}, congrArg Nat.pred h⟩
+
+/--The `tail` of `cons a v` is `v` for every vector `v` -/
+theorem tail_cons (a : α) : ∀ v : Vector α n, tail (cons a v) = v := by
+  intro v
+  match v with
+  | ⟨{data := xs}, h⟩ => rfl
+
+/--For every non-empty vector `v`, `cons (head v) (tail v)` is equal to `v`-/
+@[simp]
+theorem cons_head_tail : ∀ v : Vector α (Nat.succ n), cons (head v) (tail v) = v
+  | ⟨{data := []}, h⟩ => by contradiction
+  | ⟨{data := a :: v}, h⟩ => rfl
+
+
+/-- The array obtained from a Vector-/
+def toArray (v : Vector α n) : Array α :=
+  v.1
+
+/-- The list obtained from a vector. -/
+def toList (v : Vector α n) : List α :=
+  v.1.1
+
+
+/-- nth element of a vector, indexed by a `Fin` type. -/
+def get (v : Vector α n) (i : Fin n) : α :=
+  v.1.get <| i.cast v.2.symm
+
+/-- Appending a vector to another. -/
+def append {n m : Nat} : Vector α n → Vector α m → Vector α (n + m)
+  | ⟨a₁, h₁⟩, ⟨a₂, h₂⟩ => ⟨a₁ ++ a₂, by rw[Array.size_append, ←h₁,←h₂]⟩
+
+
+/--Eliminator for `Vector α n`-/
+@[elab_as_elim]
+def elim {α} {C : ∀ {n}, Vector α n → Sort u}
+    (H : ∀ a : Array α, C ⟨a, rfl⟩) {n : Nat} : ∀ v : Vector α n, C v
+  | ⟨a, h⟩ =>
+    match n, h with
+    | _, rfl => H a
+
+
+/-- Map a vector under a function. -/
+def map (f : α → β) : Vector α n → Vector β n
+  | ⟨{data := l}, h⟩ => ⟨{data := List.map f l}, by simp [*]⟩
+
+/-- A `nil` vector maps to a `nil` vector. -/
+@[simp]
+theorem map_nil (f : α → β) : map f nil = nil :=
+  rfl
+
+
+/-- `map` is natural with respect to `cons`. -/
+@[simp]
+theorem map_cons (f : α → β) (a : α) : ∀ v : Vector α n, map f (cons a v) = cons (f a) (map f v)
+  | ⟨{data := _}, _⟩ => rfl
+
+
+/-- Mapping two vectors under a curried function of two variables. -/
+def map₂ (f : α → β → φ) : Vector α n → Vector β n → Vector φ n
+  | ⟨a, h₁⟩, ⟨b, h₂⟩ => ⟨Array.zipWith a b f, by simp [Array.size_zipWith, h₁, h₂]⟩
+
+/-- Vector obtained by repeating an element. -/
+def replicate (n : Nat) (a : α) : Vector α n :=
+  ⟨{data := List.replicate n a}, List.length_replicate n a⟩
```

Comment:
As a user I would expect an `Array` backed vector to have efficient implementations for all operations that don't go through lists (because if I wouldn't care about performance, certainly backing this by lists or an inductive type would be convenient for proving). So operations like this ought to be backed by efficient Array operations. Where those don't exist yet, maybe they need to be added?

Reply:
This is correct. I am changing the operations one by one. However there are some operations like `replicate` that would be `O(n)` regardless. So does it make sense to replace those?

Reply:
Yes! It's not just about asymptomatics; creating a list first and then converting to an array is slower than allocating an array (of known final size even!) and filling it mutably.

Reply:
I would use `ofFn` to implement this one.

Reply:
I already replaced this

Reply:
Strictly speaking it is a duplicate of `mkVector` again present because of the Mathlib Vector API. I think it should be deleted altogether